File: update_data.py
# 파일명: update_data.py
# 경로: 깃허브 저장소 최상위 (main 브랜치 루트)

import os
import re
import json
import time
import random
import requests
from bs4 import BeautifulSoup
from datetime import datetime, timedelta

USER_AGENTS = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
]
KAKAO_API_KEY = "a6b27b6dab16c7e3459bb9589bf1269d"

def get_headers():
    return {"User-Agent": random.choice(USER_AGENTS), "Accept-Language": "ko-KR,ko;q=0.9"}

def load_js_data(filepath, is_dict):
    if not os.path.exists(filepath): return {} if is_dict else []
    with open(filepath, 'r', encoding='utf-8') as f: content = f.read()
    start = content.find('{') if is_dict else content.find('[')
    if start == -1: return {} if is_dict else []
    json_str = content[start:].strip()
    if json_str.endswith(';'): json_str = json_str[:-1]
    try: return json.loads(json_str)
    except: return {} if is_dict else []

def save_js_data(filepath, var_name, data):
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(f"const {var_name} = ")
        json.dump(data, f, ensure_ascii=False, indent=4)
        f.write(";\n")

def get_latest_round():
    try:
        first_draw_date = datetime(2002, 12, 7, 21, 0, 0)
        korea_time = datetime.utcnow() + timedelta(hours=9)
        delta = korea_time - first_draw_date
        current_round = int(delta.total_seconds() // (7 * 24 * 3600)) + 1
        return current_round
    except:
        return 0

def fetch_winning_numbers(draw_no):
    # [핵심 수정] 찾아주신 레퍼런스와 100% 동일한 방식: 동행복권 HTML 결과 페이지 직접 스크래핑
    url = f"https://dhlottery.co.kr/gameResult.do?method=byWin&drwNo={draw_no}"
    try:
        r = requests.get(url, headers=get_headers(), timeout=10)
        soup = BeautifulSoup(r.text, "html.parser")

        # 레퍼런스의 태그 셀렉터(div.num.win span.ball_645) 그대로 적용
        balls = soup.select('div.num.win span.ball_645')
        if len(balls) == 6:
            nums = [int(ball.text.strip()) for ball in balls]
            
            # 보너스 번호 태그 셀렉터 적용
            bonus_ball = soup.select_one('div.num.bonus span.ball_645')
            if bonus_ball:
                nums.append(int(bonus_ball.text.strip()))
                return nums
    except: pass
    return None

def fetch_stores_news(draw_no):
    stores_map = {}
    queries = [f"로또 {draw_no}회 1등 배출점", f"로또 {draw_no}회 1등 판매점"]
    links = []
    
    # 1. 네이버 뉴스 우선 탐색
    for query in queries:
        url = f"https://search.naver.com/search.naver?where=news&query={requests.utils.quote(query)}&sort=1"
        try:
            r = requests.get(url, headers=get_headers(), timeout=5)
            soup = BeautifulSoup(r.text, "html.parser")
            for a in soup.find_all('a', href=True):
                href = a['href']
                if ('news.naver.com' in href or 'n.news.naver.com' in href) and 'article' in href:
                    if href not in links: links.append(href)
            if len(links) >= 3: break
        except: continue

    # 2. 다음 뉴스 탐색 (네이버가 차단했을 때 백업)
    if not links:
        for query in queries:
            url = f"https://search.daum.net/search?w=news&q={requests.utils.quote(query)}"
            try:
                r = requests.get(url, headers=get_headers(), timeout=5)
                soup = BeautifulSoup(r.text, "html.parser")
                for a in soup.find_all('a', href=True):
                    href = a['href']
                    if 'v.daum.net/v/' in href:
                        if href not in links: links.append(href)
                if len(links) >= 3: break
            except: continue

    if not links:
        print("  ⚠️ 포털(네이버/다음) 뉴스에서 1등 배출점 관련 기사를 찾지 못했습니다.")
        return []

    blacklist = ["홈페이지", "뉴스", "기자", "기사", "동행복권", "인터넷", "판매점", "당첨", "연합뉴스"]
    for link in links[:5]:
        try:
            nr = requests.get(link, headers=get_headers(), timeout=5)
            nr.encoding = 'utf-8'
            nsoup = BeautifulSoup(nr.text, "html.parser")
            
            content = nsoup.find('article') or nsoup.find('div', id='dic_area') or nsoup.find('div', class_='article_view')
            if not content: continue
            
            text = content.get_text(separator="\n")
            matches = list(re.finditer(r'(?:▲|△|■|▶|●|\*|\-|[\d]+\.)\s*([가-힣\w\d&/\s()+\-.,]+?)\s*\(([^)]+)\)', text))
            for m in matches:
                p1, p2 = m.group(1).strip(), m.group(2).strip()
                if any(word in p1 for word in blacklist) and len(p1) < 10: continue
                if len(p1) < 2 or len(p2) < 2: continue

                if any(x in p1 for x in ['시 ', '구 ', '군 ', '읍 ', '면 ', '리 ']) or re.search(r'\d+-\d+', p1):
                    name, addr = p2, p1
                else:
                    name, addr = p1, p2

                method = "자동"
                if "수동" in text[max(0, m.start()-80):m.end()+150]: method = "수동"
                if "반자동" in text[max(0, m.start()-80):m.end()+150]: method = "반자동"

                key = (name, addr)
                if key not in stores_map: stores_map[key] = method
        except: continue

    return [{"n": n, "a": a, "m": m, "r": draw_no} for (n, a), m in stores_map.items()]

def geocode(address):
    addr = re.sub(r"(\d+)억?", r"\1", address).rstrip("., ").strip()
    try:
        r = requests.get("https://dapi.kakao.com/v2/local/search/address.json", headers={"Authorization": "KakaoAK " + KAKAO_API_KEY}, params={"query": addr}, timeout=5)
        data = r.json()
        if data.get("documents"): return float(data["documents"][0]["y"]), float(data["documents"][0]["x"])
    except: pass
    return 0.0, 0.0

def main():
    history_file = 'lotto_history_data.js'
    data_file = 'lotto_data.js'

    latest_round = get_latest_round()
    if latest_round == 0: 
        print("❌ 회차 계산에 실패했습니다.")
        return

    print(f"✅ 현재 최신 회차: {latest_round}회")
    print("🔍 최근 5회차의 누락분만 빠르게 스캔합니다.\n")

    hist_data = load_js_data(history_file, is_dict=True)
    store_data = load_js_data(data_file, is_dict=False)
    existing_store_rounds = set(item.get('r') for item in store_data if isinstance(item, dict) and 'r' in item)

    unsaved_updates = 0
    
    start_round = max(1, latest_round - 5)

    for draw_no in range(start_round, latest_round + 1):
        missing_hist = str(draw_no) not in hist_data
        missing_store = draw_no not in existing_store_rounds

        if not missing_hist and not missing_store:
            print(f"⏩ [{draw_no}회차] 데이터 정상 보유 (패스)")
            continue

        print(f"▶ [{draw_no}회차] 누락 확인. 데이터 수집 시작...")

        if missing_hist:
            nums = fetch_winning_numbers(draw_no)
            if nums:
                hist_data[str(draw_no)] = nums
                print(f"  ✅ 당첨번호 복구 완료: {nums}")
                unsaved_updates += 1
            else:
                print(f"  ❌ 당첨번호 수집 실패 (웹페이지 크롤링 차단됨)")

        if missing_store:
            stores = fetch_stores_news(draw_no)
            if stores:
                for st in stores:
                    is_online = "인터넷" in st["n"] or "사이트" in st["m"] or "dhlottery" in st["n"].lower()
                    if is_online:
                        st["n"], st["a"] = "동행복권(dhlottery.co.kr)", "서울특별시 서초구 남부순환로 2423 1층"
                        lat, lng = 37.4831, 127.0225
                    else:
                        lat, lng = geocode(st["a"])

                    store_data.append({"r": draw_no, "n": st["n"], "m": st["m"], "a": st["a"], "lat": lat, "lng": lng, "verified": is_online})
                print(f"  ✅ 1등 판매점 {len(stores)}곳 복구 완료")
                unsaved_updates += 1

        # [핵심 추가] 서버 부담 완화 및 차단 방지를 위한 1초 대기 (마지막 회차 제외)
        if draw_no < latest_round:
            time.sleep(1)

    if unsaved_updates > 0:
        store_data.sort(key=lambda x: x.get('r', 0) if isinstance(x, dict) else 0, reverse=True)
        save_js_data(history_file, 'LOTTO_HISTORY', hist_data)
        save_js_data(data_file, 'lottoData', store_data)
        print("\n💾 새로운 데이터가 파일에 안전하게 기록되었습니다.")
    else:
        print("\n🎉 모든 데이터가 최신 상태입니다. 업데이트할 내용이 없습니다.")

if __name__ == "__main__":
    main()

File: update_data.js
// 파일명: update_data.js
// 경로: 깃허브 저장소 최상위 (main 브랜치 루트)

const fs = require('fs');
const axios = require('axios');
const cheerio = require('cheerio');

const USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36';
const KAKAO_API_KEY = "a6b27b6dab16c7e3459bb9589bf1269d";

function getLatestRound() {
    const firstDrawDate = new Date('2002-12-07T21:00:00+09:00');
    const now = new Date();
    const diff = now.getTime() - firstDrawDate.getTime();
    if (diff < 0) return 0;
    return Math.floor(diff / (7 * 24 * 60 * 60 * 1000)) + 1;
}

function loadJsData(filepath, isDict) {
    if (!fs.existsSync(filepath)) return isDict ? {} : [];
    let content = fs.readFileSync(filepath, 'utf8');
    let start = isDict ? content.indexOf('{') : content.indexOf('[');
    if (start === -1) return isDict ? {} : [];
    let jsonStr = content.substring(start).trim();
    if (jsonStr.endsWith(';')) jsonStr = jsonStr.slice(0, -1);
    try { return JSON.parse(jsonStr); } catch (e) { return isDict ? {} : []; }
}

function saveJsData(filepath, varName, data) {
    const jsonStr = JSON.stringify(data, null, 4);
    fs.writeFileSync(filepath, `const ${varName} = ${jsonStr};\n`, 'utf8');
}

// [핵심 변경] 깃허브 IP가 동행복권에서 차단될 경우를 완벽하게 대비하여
// 동행복권 웹 ➔ 동행복권 API ➔ 다음(Daum) 검색 ➔ 네이버 검색 순으로 4중 백업 파싱 로직 적용
async function fetchWinningNumbers(drawNo) {
    // 1순위: 동행복권 웹페이지
    try {
        const res = await axios.get(`https://dhlottery.co.kr/gameResult.do?method=byWin&drwNo=${drawNo}`, { headers: { 'User-Agent': USER_AGENT }, timeout: 5000 });
        const $ = cheerio.load(res.data);
        const numbers = [];
        $('div.num.win span.ball_645').each((i, el) => numbers.push(parseInt($(el).text().trim(), 10)));
        if (numbers.length === 6) {
            const bonusStr = $('div.num.bonus span.ball_645').text().trim();
            if (bonusStr) { numbers.push(parseInt(bonusStr, 10)); return numbers; }
        }
    } catch (e) { console.log(`  [동행복권 웹 차단됨] ${e.message}`); }

    // 2순위: 동행복권 API
    try {
        const res = await axios.get(`https://www.dhlottery.co.kr/common.do?method=getLottoNumber&drwNo=${drawNo}`, { headers: { 'User-Agent': USER_AGENT }, timeout: 5000 });
        if (res.data && res.data.returnValue === "success") {
            const nums = [];
            for (let i = 1; i <= 6; i++) nums.push(res.data[`drwtNo${i}`]);
            nums.push(res.data.bnusNo);
            return nums;
        }
    } catch (e) { console.log(`  [동행복권 API 차단됨] ${e.message}`); }

    // 3순위: 다음(Daum) 검색 (해외 IP 차단이 없어 깃허브 액션에서 가장 안전하게 동작)
    try {
        const res = await axios.get(`https://search.daum.net/search?w=tot&q=${encodeURIComponent(`로또 ${drawNo}회 당첨번호`)}`, { headers: { 'User-Agent': USER_AGENT }, timeout: 5000 });
        const text = res.data.replace(/<[^>]*>?/gm, ' ');
        const mainMatch = text.match(/(?:당첨번호|당첨 번호)\D{0,15}?(\d{1,2})\D{1,5}?(\d{1,2})\D{1,5}?(\d{1,2})\D{1,5}?(\d{1,2})\D{1,5}?(\d{1,2})\D{1,5}?(\d{1,2})/);
        if (mainMatch) {
            const nums = mainMatch.slice(1, 7).map(n => parseInt(n, 10));
            const bonusMatch = text.match(/보너스\D{0,15}?(\d{1,2})(?!\d|등)/);
            if (bonusMatch) {
                const bonus = parseInt(bonusMatch[1], 10);
                if (!nums.includes(bonus)) { nums.push(bonus); return nums; }
            }
        }
    } catch (e) { console.log(`  [Daum 검색 에러] ${e.message}`); }

    // 4순위: 네이버 검색
    try {
        const res = await axios.get(`https://search.naver.com/search.naver?query=${encodeURIComponent(`로또 ${drawNo}회 당첨번호`)}`, { headers: { 'User-Agent': USER_AGENT }, timeout: 5000 });
        const text = res.data.replace(/<[^>]*>?/gm, ' ');
        const mainMatch = text.match(/(?:당첨번호|당첨 번호)\D{0,15}?(\d{1,2})\D{1,5}?(\d{1,2})\D{1,5}?(\d{1,2})\D{1,5}?(\d{1,2})\D{1,5}?(\d{1,2})\D{1,5}?(\d{1,2})/);
        if (mainMatch) {
            const nums = mainMatch.slice(1, 7).map(n => parseInt(n, 10));
            const bonusMatch = text.match(/보너스\D{0,15}?(\d{1,2})(?!\d|등)/);
            if (bonusMatch) {
                const bonus = parseInt(bonusMatch[1], 10);
                if (!nums.includes(bonus)) { nums.push(bonus); return nums; }
            }
        }
    } catch (e) { console.log(`  [Naver 검색 에러] ${e.message}`); }

    return null;
}

// [수정됨] 1등 배출점 또한 동행복권 차단 시 포털 뉴스로 즉각 우회하는 백업 로직 통합
async function fetchStores(drawNo) {
    let stores = [];
    try {
        const res = await axios.get(`https://dhlottery.co.kr/store.do?method=topStore&pageGubun=L645&drwNo=${drawNo}`, { headers: { 'User-Agent': USER_AGENT }, timeout: 5000 });
        const $ = cheerio.load(res.data);
        $('table.tbl_data').eq(0).find('tbody tr').each((i, el) => {
            const tds = $(el).find('td');
            if (tds.length >= 4) {
                const name = $(tds[1]).text().trim();
                const method = $(tds[2]).text().trim();
                const addr = $(tds[3]).text().trim();
                if (name && !name.includes('조회 결과가 없습니다')) {
                    stores.push({ n: name, m: method, a: addr, r: drawNo });
                }
            }
        });
        if (stores.length > 0) return stores;
    } catch (e) { console.log(`  [동행복권 배출점 차단됨] ${e.message}`); }

    console.log(`  ⚠️ 동행복권 직접 수집 불가. 포털 뉴스를 통해 우회 수집합니다.`);
    return await fetchStoresFromNews(drawNo);
}

async function fetchStoresFromNews(drawNo) {
    const storesMap = new Map();
    const queries = [`로또 ${drawNo}회 1등 배출점`, `로또 ${drawNo}회 1등 판매점`];
    const links = [];

    for (const q of queries) {
        try {
            const res = await axios.get(`https://search.naver.com/search.naver?where=news&query=${encodeURIComponent(q)}&sort=1`, { headers: { 'User-Agent': USER_AGENT }, timeout: 5000 });
            const $ = cheerio.load(res.data);
            $('a').each((i, el) => {
                const href = $(el).attr('href');
                if (href && (href.includes('news.naver.com') || href.includes('n.news.naver.com')) && href.includes('article')) {
                    if (!links.includes(href)) links.push(href);
                }
            });
            if (links.length >= 3) break;
        } catch(e){}
    }

    if (links.length === 0) {
        for (const q of queries) {
            try {
                const res = await axios.get(`https://search.daum.net/search?w=news&q=${encodeURIComponent(q)}`, { headers: { 'User-Agent': USER_AGENT }, timeout: 5000 });
                const $ = cheerio.load(res.data);
                $('a').each((i, el) => {
                    const href = $(el).attr('href');
                    if (href && href.includes('v.daum.net/v/')) {
                        if (!links.includes(href)) links.push(href);
                    }
                });
                if (links.length >= 3) break;
            } catch(e){}
        }
    }

    const blacklist = ["홈페이지", "뉴스", "기자", "기사", "동행복권", "인터넷", "판매점", "당첨", "연합뉴스"];
    for (const link of links.slice(0, 5)) {
        try {
            const res = await axios.get(link, { headers: { 'User-Agent': USER_AGENT }, timeout: 5000 });
            const $ = cheerio.load(res.data);
            const article = $('article').text() || $('#dic_area').text() || $('.article_view').text();
            if (!article) continue;

            const regex = /(?:▲|△|■|▶|●|\*|\-|[\d]+\.)\s*([가-힣\w\d&\/\s()+\-.,]+?)\s*\(([^)]+)\)/g;
            let match;
            while ((match = regex.exec(article)) !== null) {
                let p1 = match[1].trim();
                let p2 = match[2].trim();

                if (blacklist.some(b => p1.includes(b)) && p1.length < 10) continue;
                if (p1.length < 2 || p2.length < 2) continue;

                let name, addr;
                if (['시 ', '구 ', '군 ', '읍 ', '면 ', '리 '].some(x => p1.includes(x)) || /\d+-\d+/.test(p1)) {
                    name = p2; addr = p1;
                } else {
                    name = p1; addr = p2;
                }

                let method = "자동";
                const context = article.substring(Math.max(0, match.index - 80), match.index + 150);
                if (context.includes("수동")) method = "수동";
                if (context.includes("반자동")) method = "반자동";

                const key = `${name}|${addr}`;
                if (!storesMap.has(key)) {
                    storesMap.set(key, { n: name, m: method, a: addr, r: drawNo });
                }
            }
        } catch(e){}
    }
    return Array.from(storesMap.values());
}

async function geocode(address) {
    let addr = address.replace(/(\d+)억?/g, '$1').replace(/[.,\s]+$/, '').trim();
    try {
        const url = `https://dapi.kakao.com/v2/local/search/address.json?query=${encodeURIComponent(addr)}`;
        const res = await axios.get(url, { headers: { 'Authorization': `KakaoAK ${KAKAO_API_KEY}` }, timeout: 5000 });
        const docs = res.data.documents;
        if (docs && docs.length > 0) {
            return { lat: parseFloat(docs[0].y), lng: parseFloat(docs[0].x) };
        }
    } catch (e) { }
    return { lat: 0.0, lng: 0.0 };
}

async function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function main() {
    const historyFile = 'lotto_history_data.js';
    const dataFile = 'lotto_data.js';

    const latestRound = getLatestRound();
    if (latestRound === 0) return;

    console.log(`✅ 현재 최신 회차: ${latestRound}회`);

    const histData = loadJsData(historyFile, true);
    const storeData = loadJsData(dataFile, false);
    const existingStoreRounds = new Set(storeData.map(item => item.r));

    let unsavedUpdates = 0;
    const startRound = Math.max(1, latestRound - 5);

    for (let drawNo = startRound; drawNo <= latestRound; drawNo++) {
        const missingHist = !(drawNo.toString() in histData);
        const missingStore = !existingStoreRounds.has(drawNo);

        if (!missingHist && !missingStore) {
            console.log(`⏩ [${drawNo}회차] 데이터 정상 보유 (패스)`);
            continue;
        }

        console.log(`▶ [${drawNo}회차] 누락 확인. 데이터 수집 시작...`);

        if (missingHist) {
            const nums = await fetchWinningNumbers(drawNo);
            if (nums) {
                histData[drawNo.toString()] = nums;
                console.log(`  ✅ 당첨번호 복구 완료: ${nums}`);
                unsavedUpdates++;
            } else {
                console.log(`  ❌ 당첨번호 수집 실패 (모든 우회 경로 막힘)`);
            }
        }

        if (missingStore) {
            const stores = await fetchStores(drawNo);
            if (stores && stores.length > 0) {
                for (let st of stores) {
                    const isOnline = st.n.includes("인터넷") || st.m.includes("사이트") || st.n.toLowerCase().includes("dhlottery");
                    let lat = 0.0, lng = 0.0;
                    if (isOnline) {
                        st.n = "동행복권(dhlottery.co.kr)";
                        st.a = "서울특별시 서초구 남부순환로 2423 1층";
                        lat = 37.4831;
                        lng = 127.0225;
                    } else {
                        const coords = await geocode(st.a);
                        lat = coords.lat;
                        lng = coords.lng;
                    }
                    storeData.push({ r: drawNo, n: st.n, m: st.m, a: st.a, lat: lat, lng: lng, verified: isOnline });
                }
                console.log(`  ✅ 1등 판매점 ${stores.length}곳 복구 완료`);
                unsavedUpdates++;
            } else {
                console.log(`  ⚠️ 1등 배출점 정보가 아직 없거나 모두 차단되었습니다.`);
            }
        }

        if (drawNo < latestRound) {
            await sleep(1000);
        }
    }

    if (unsavedUpdates > 0) {
        storeData.sort((a, b) => b.r - a.r);
        saveJsData(historyFile, 'LOTTO_HISTORY', histData);
        saveJsData(dataFile, 'lottoData', storeData);
        console.log(`\n💾 새로운 데이터 기록 완료.`);
    } else {
        console.log(`\n🎉 업데이트할 내용이 없습니다.`);
    }
}

main().catch(err => console.error(err));
